### ISSUE TITLE
Dialogs in macOS don't close

### ISSUE THREAD

**Issue by Grix** [time-killer-games/libdlgmod#5]
Hi again. I have finally been attempting to update to the newest version of the library without the SDL2 dependence, but I am seeing some big bugs on macOS.  

Message and question dialogs doesn't disappear when I click their buttons. This is the synchronous dialogs, haven't tested async. 

It also doesn't looks like the save file dialogs properly return a path either, because no file is created, but this might be because of other issues. It's hard to debug since I cannot run my program in GM debug mode on mac since it crashes due to library code signing shenanigans. So let's focus on the first issue first.

This is very similar (or maybe identical) to issues with a certain build we discussed earlier in another issue here: https://github.com/time-killer-games/libdlgmod/issues/3#issuecomment-2052163282

That time, you linked a project which you had confirmed to work, and indeed it worked for me too. So I left it at that since I wasn't planning on actually update the library myself yet. However, now that I am trying to update, the issue is somehow back, so probably we should get to the bottom of it. The link to the working project you posted back then is dead, and also it seems the github releases aren't archived, I only see the latest, which is not working. So I cannot easily compare the difference between the working and non-working versions. 

I tried both the precompiled binaries in the github release, and building myself with build.sh, same result for both.

In Windows it is working fine.

**Comment by time-killer-games** [comment]
I'm on vacation right now and don't have access to my macOS machine. I'll be heading back home January 2nd. I'll let you know when I get back when I am ready to work with you on this.

There a project I've contributed to which actively uses both the save dialogs and the message/question dialogs on macOS called Minecraft Note Block Studio and that project is not experiencing these issues you mentioned. I wonder what is causing it to do these things for you.

**Comment by time-killer-games** [comment]
I'm on vacation right now and don't have access to my macOS machine. I'll be heading back home January 2nd. I'll let you know when I get back when I am ready to work with you on this.

There a project I've contributed to which actively uses both the save dialogs and the message/question dialogs on macOS called Minecraft Note Block Studio and that project is not experiencing these issues you mentioned. I wonder what is causing it to do these things for you.

Edit:

I'm not sure if this will make any difference, and I doubt this I related to your problem at all, but do you set a parent/owner window for the dialogs in your project? This is how you do it, and this function must be called before showing any of the dialogs:

```
var wid = string(int64(window_handle()));
widget_set_owner(wid);
```

**Comment by Grix** [comment]
Wow, success! I added the set-owner code you gave and it seems to have fixed everything, indeed. At least I think that's what did it. Thank you

I will also try moving to async dialogs per your recommendations. 

I will keep this issue open until I can do even more thorough testing, if you don't mind, but it looks very promising.

**Comment by time-killer-games** [comment]
> Wow, success! I added the set-owner code you gave and it seems to have fixed everything, indeed. At least I think that's what did it, I don't think I made any other relevant changes since last time I tried. Thank you
> 
> I will also try moving to async dialogs per your recommendations. 
> 
> I will keep this issue open until I can do even more thorough testing, if you don't mind, but it looks very promising. 

Yeah, I had a very strange feeling that setting the owner window would do the trick. I recalled when no owner was set it had some strange behaviors, but I couldn't remember what the exact problems were.

I'm so glad that it looks like we figured it out! I hope you enjoy the extension a lot more now that it is native and not custom drawn with Dear ImGui. It's a lot better.

**Comment by Grix** [comment]
New Mac dialogs seem to be working well, so I'm closing this issue (although as you see I made another issue for a separate problem :) )
